components/support/android/src/main/java/mozilla/appservices/support/SyncTelemetryPing.kt

```diff
@@ -0,0 +1,248 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```

Comment:
how do you picture the glean integration working? I guess I was kinda assuming that the glean integration part might just do the string -> json -> glean in the same module and avoid exposing this to the app layer, but that's probably because I don't understand how that integration will shape up.

Reply:
IIUC, the integration needs to happen at the app layer, that's part of the problem. 😕  Fenix defines the schema for the pings ([YAML files](https://github.com/linacambridge/fenix/commit/9145a0ab7c3746181ff353dcc8928535e1b5725f) that are code-generated into Kotlin bindings), so the ping needs to get passed through Android Components. a-c can't directly translate this into Glean, because it provides Glean as a "service", not a "feature", and it doesn't define the actual metric names.

We _could_ have Android Components pass JSON strings around, and have Fenix deserialize them, but I don't think they'd be too keen on that. Their wrappers around our APIs ([`Connection::syncHistory` and `Connection::syncBookmarks`](https://github.com/mozilla-mobile/android-components/blob/a455e9c300b4d5fbf57f356628f4bc5b8cb4450b/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Connection.kt#L25-L33)) would need to return strings, we'd need to pass those opaque strings up through the different layers to Fenix, and Fenix would need to know that it's a JSON telemetry ping that it should unpack and send.

So the way this works is:

* We deserialize the ping from Rust into a `mozilla.appservices.support.SyncTelemetryPing`.
* Android Components passes that up through [all its layers](https://github.com/linacambridge/android-components/commit/579391e7937d4d6f71341b9dd195569b4fcd17d4) to a new `onStoreSynced()` observer, which takes the ping as the argument. (There's some [additional serialization](https://github.com/linacambridge/android-components/commit/bab3ca1a3965107c851113d97acedfdada3432ce) we have to do here, since it also runs the sync in a worker. See mozilla-mobile/android-components#2971).
* Fenix unpacks the `SyncTelemetryPing`, and calls the Glean submission methods like [this](https://github.com/linacambridge/fenix/commit/9145a0ab7c3746181ff353dcc8928535e1b5725f).

@grigoryk, does that make sense to you? Are there ways we can simplify?

Reply:
(Also, I'm planning to add a comment to this file explaining how this all works, and why! 😄)